src/dag.rs

```diff
@@ -599,52 +627,54 @@ impl<D: DagLike, S: SharingTracker<D>> Iterator for PostOrderIter<D, S> {
+                    //
+                    // Check whether this is the case, and record the item's new index if not.
+                    let current_index = self.tracker.record(
+                        &current.elem,
+                        self.index,
+                        current.left_idx,
+                        current.right_idx,
+                    );
+                    let already_yielded = current_index.is_some();
+                    let current_index = current_index.unwrap_or(self.index);
+
+                    // Then, update the item's parents and/or sibling, to make sure that its
+                    // parent has the correct child index. We need to do this whether or not
+                    // we're going to skip the element.
                     let stack_len = self.stack.len();
                     match current.previous {
                         Previous::Root => {
                             assert_eq!(stack_len, 0);
                         }
                         Previous::ParentLeft => {
                             assert!(self.stack[stack_len - 1].processed);
-                            self.stack[stack_len - 1].left_idx = Some(self.index);
+                            self.stack[stack_len - 1].left_idx = Some(current_index);
                         }
                         Previous::ParentRight => {
                             assert!(self.stack[stack_len - 1].processed);
-                            self.stack[stack_len - 1].right_idx = Some(self.index);
+                            self.stack[stack_len - 1].right_idx = Some(current_index);
                         }
                         Previous::SiblingLeft => {
                             assert!(self.stack[stack_len - 2].processed);
-                            self.stack[stack_len - 2].left_idx = Some(self.index);
-                            check_sibling = true;
+                            self.stack[stack_len - 2].left_idx = Some(current_index);
                         }
                     }
-                    self.tracker.record(
-                        &current.elem,
-                        self.index,
-                        current.left_idx,
-                        current.right_idx,
-                    );
-                    // There is a special case here where we are yielding a node which
-                    // is identical to its sibling (according to the sharing tracker).
-                    // In this case we have a repeated node which nonetheless made it
-                    // into the stack, so we pop it here.
-                    if check_sibling
-                        && self
-                            .tracker
-                            .seen_before(&self.stack[stack_len - 1].elem)
-                            .is_some()
-                    {
-                        self.stack[stack_len - 2].right_idx = Some(self.index);
-                        self.stack.pop();
```

Comment:
5364f815565567e73abbda4a4d03aa41944bc84e: Regarding `self.stack.pop()`: This line pops the sibling on `master`, which is removed by the commit.

The current item is already popped at the top of the loop, you're right, but its sibling still sits on the stack unless we pop it. We assert that the sibling is processed, so it is popped and skipped by the next loop iteration.

Reply:
Right, exactly. It was wrong to treat siblings as being special; we could have repeats of siblings *or* any of their descendants, which requires more general code.